Custom Factors in Python

ProfFan opened the pull request:
Per #748 and #764, this is a factor that allow one to create the evaluateError function in Python.

The unit tests document the usage.

It also contains a fix for the proper binding of `Point3Pairs`. @johnwlambert

dellaert commented on gtsam/nonlinear/CustomFactor.cpp:
```diff
@@ -0,0 +1,35 @@
+Vector CustomFactor::unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H) const {
+  if(this->active(x)) {
+    if(H) {
+      return this->errorFunction(*this, x, H.get_ptr());
```
Document what the mechanism is by which python can access. I expect you'll talk about pybind, types, how it appears in python, what python function should do.

ProfFan replied:
done

dellaert commented on gtsam/nonlinear/CustomFactor.cpp:
```diff
@@ -0,0 +1,35 @@
+      return this->errorFunction(*this, x, H.get_ptr());
+    } else {
+      JacobianVector dummy;
+      return this->errorFunction(*this, x, &dummy);
```
Document what happens in  this case.

ProfFan replied:
done

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,86 @@
+from gtsam.utils.test_case import GtsamTestCase
+
+
+class TestCustomFactor(GtsamTestCase):
```
doc

ProfFan replied:
done

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,86 @@
+
+class TestCustomFactor(GtsamTestCase):
+
+    def test_new(self):
```
doc! Don't you use a linter?

ProfFan replied:
done

dellaert reviewed (changes requested):
You **have to** document this. You are the *only* one* who will do this. It has to also show up in some markdown file, dson't know which one. This is probably worth thinking about re the "readthedocs" change.

gchenfc commented on matlab/CMakeLists.txt:
```diff
@@ -61,7 +61,8 @@ endif()
     gtsam::Point2
-    gtsam::Point3)
+    gtsam::Point3
+    gtsam::CustomFactor)
```
Can you add a comment in gtsam.i that CustomFactor won't be wrapped for matlab?

ProfFan replied:
done

gchenfc commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,86 @@
+            return np.array([1, 0, 0])
+        
+        noise_model = gtsam.noiseModel.Unit.Create(3)
+        cf = CustomFactor(noise_model, gtsam.KeyVector([0]), error_func)
```
is `gtsam.KeyVector(...)` necessary?  I thought for KeyVector then

```python
cf = CustomFactor(noise_model, [0], error_func)
```

should work?  If not, is there a different type for which this would work?

varunagrawal replied:
We should consider making KeyVector default to a regular list on the python side while keeping the allocator semantics on the cpp side

ProfFan replied:
Now it should work, I added the implicit conversion for `list`.

gchenfc commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,86 @@
+            gT1, gT2 = v.atPose2(key0), v.atPose2(key1)
+            error = Pose2(0, 0, 0).localCoordinates(gT1.between(gT2))
+            
+            if len(H) > 0:
```
I feel the pythonic syntax would be `if H is not None: `, is there any way we can achieve this?  I know one way would be to include pybind in `CustomFactor.h` then use `pybind::none`, but I guess this isn't ideal.  Hoping for a null_ptr way or something.

varunagrawal replied:
Yeah I want this too. I'll take a look at potential fixes later tonight. Lots of stuff to do for internship.

ProfFan replied:
This is a great idea. I just switched to your suggestion :)

gchenfc commented on gtsam/nonlinear/CustomFactor.cpp:
```diff
@@ -0,0 +1,35 @@
+    if(H) {
+      return this->errorFunction(*this, x, H.get_ptr());
+    } else {
+      JacobianVector dummy;
+      return this->errorFunction(*this, x, &dummy);
```
could this pass in null_ptr instead or something?  Possibly by something along the lines of

```c++
typedef std::function<Vector(const CustomFactor&, const Values&, std::optional<JacobianVector*>)> CustomErrorFunction;
```
or something ???

varunagrawal replied:
That's apparently doable as per some issues on Pybind11, but we'd have to use boost::optional because cpp11

ProfFan replied:
Yes, done :)

gchenfc commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,84 @@
+
+class CustomFactor;
+
+typedef std::function<Vector(const CustomFactor&, const Values&, const JacobianVector*)> CustomErrorFunction;
```
Reason for `JacobianVector *` to be const?

ProfFan replied:
It's a const pointer to an array (whose size is preallocated by GTSAM). 

Design note: pybind11 will invoke a copy if this is `JacobianVector &`, and modifications in Python will not be reflected.

I'll add this to the comments

ProfFan replied:
done

gchenfc commented on gtsam/gtsam.i:
```diff
@@ -2166,6 +2166,12 @@ virtual class NoiseModelFactor: gtsam::NonlinearFactor {
+#include <gtsam/nonlinear/CustomFactor.h>
+virtual class CustomFactor: gtsam::NoiseModelFactor {
+  CustomFactor();
+  CustomFactor(const gtsam::SharedNoiseModel& noiseModel, const gtsam::KeyVector& keys, const gtsam::CustomErrorFunction& errorFunction);
```
can you add a documentation/comment here about the args of `CustomErrorFunction`, just so that anyone using gtsam.i as an API reference doesn't have to go digging around to figure out how to use it?

varunagrawal replied:
This should be add-able via the docstring mechanism. That way, the editor shows it. Adding usage docs should go in the cpp file and not in the .i file imo

ProfFan replied:
done

gchenfc reviewed (approved):
Cool!

See (minor) comments

johnwlambert commented on python/gtsam/preamble.h:
```diff
@@ -6,9 +6,17 @@ PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Key, tbb::tbb_allocator<gtsam::Key>>);
 PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Key>);
 #endif
 PYBIND11_MAKE_OPAQUE(std::vector<gtsam::Point2, Eigen::aligned_allocator<gtsam::Point2> >);
+PYBIND11_MAKE_OPAQUE(gtsam::Point3Pairs);
```
thanks for the PR, Fan. Was this opaque option something we forgot to add earlier? What's the advantage of it -- preventing a deep copy?

ProfFan replied:
Yes, it prevents a deep copy.

varunagrawal replied:
This tells Pybind11 not to perform a cpp-to-python conversion of types and instead treat this as its own type. This help prevents issues with pass-by-reference propagation between the two languages.

ProfFan replied:
^ much better explanation

dellaert replied:
Make sure this much better explanation is copied to wrapper docs.

ProfFan replied:
done

varunagrawal commented on gtsam/gtsam.i:
````diff
@@ -2166,6 +2166,28 @@ virtual class NoiseModelFactor: gtsam::NonlinearFactor {
+   * cf = CustomFactor(noise_model, keys, error_func)
+   * ```
+   */
+  CustomFactor(const gtsam::SharedNoiseModel& noiseModel, const gtsam::KeyVector& keys, const gtsam::CustomErrorFunction& errorFunction);
````
Format just this block?

varunagrawal commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,92 @@
+ * This is safe because this is passing a const pointer, and pybind11 will maintain the `std::vector` memory layout.
+ * Thus the pointer will never be invalidated.
+ */
+typedef std::function<Vector(const CustomFactor&, const Values&, const JacobianVector*)> CustomErrorFunction;
```
use `using` instead of `typedef`?

ProfFan replied:
done

varunagrawal commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,92 @@
+
+protected:
+
+  typedef NoiseModelFactor Base;
```
`using` instead of `typedef` for `Base` and `This`.

ProfFan replied:
I think all our old factors use typedef, so let's keep using typedef here?

varunagrawal replied:
Yes but that's legacy. @dellaert mentioned to me that when we update/write new code, we should use the modern conventions. I can push changes to this PR if you're okay with that. :slightly_smiling_face:

dellaert replied:
Please use "using". @varunagrawal let's keep this PR Fan's task :-)

varunagrawal commented on python/gtsam/specializations.h:
```diff
@@ -2,9 +2,12 @@
+py::implicitly_convertible<py::list, std::vector<gtsam::Key, tbb::tbb_allocator<gtsam::Key> > >();
 #else
 py::bind_vector<std::vector<gtsam::Key> >(m_, "KeyVector");
+py::implicitly_convertible<py::list, std::vector<gtsam::Key> >();
```
TIL something new! Awesome.

varunagrawal commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,125 @@
+            return np.array([1, 0, 0])
+        
+        noise_model = gtsam.noiseModel.Unit.Create(3)
+        cf = CustomFactor(noise_model, gtsam.KeyVector([0]), error_func)
```
Update the test to use the `list` or add another one to show both options?

ProfFan replied:
Sure

varunagrawal commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,92 @@
+    * One can check if a derivative is needed in the errorFunction by checking the length of Jacobian array
+  */
+  Vector unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H = boost::none) const override;
+
```
Formatting, and also maybe override the `print` and `operator<<` methods?

ProfFan replied:
done

varunagrawal reviewed (approved):
Looks good overall, some comments and needs formatting. Awesome job @ProfFan!

varunagrawal commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,125 @@
+            gT1, gT2 = v.atPose2(key0), v.atPose2(key1)
+            error = Pose2(0, 0, 0).localCoordinates(gT1.between(gT2))
+            
+            if not H is None:
```
`if H is not None` is the pythonic way. :slightly_smiling_face:

varunagrawal commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,92 @@
+
+namespace gtsam {
+
+typedef std::vector<Matrix> JacobianVector;
```
`using` instead of `typedef`

varunagrawal commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,92 @@
+ */
+class CustomFactor: public NoiseModelFactor {
+public:
+   CustomErrorFunction errorFunction;
```
Should we make this `private` so that the error function passed in from python doesn't need `this` to be the first argument?

ProfFan replied:
I think this is not relevant, since the callback will need the `this` pointer because one can reuse the same callback for multiple factors.

varunagrawal replied:
Wouldn't we then just pass the same function to different factors. I don't think I understand but would one factor refer to another factor's errorFunction explicitly? That kinda seems more error prone to me (e.g. we are indexing a factor list and we accidentally index the wrong factor).

dellaert replied:
I don't understand either why it's public, and why it does not obey naming convention (trailing underscore).

ProfFan replied:
fixed

varunagrawal reviewed (approved):
Some more stuff I noticed and should have added in my last review.

varunagrawal commented:
@ProfFan can you please re-request a review after you make the changes? Overall it looks great but I want to do a thorough review again so I can start using this ASAP.

ProfFan commented:
I am thinking about adding an optional constructor param that allows one to specify the name of the custom factor (stored as a string), so people can distinguish different `CustomFactor` s.

dellaert reviewed (changes requested):
MArkdown?

dellaert commented on gtsam/gtsam.i:
```diff
@@ -2166,6 +2166,31 @@ virtual class NoiseModelFactor: gtsam::NonlinearFactor {
 
+#include <gtsam/nonlinear/CustomFactor.h>
+virtual class CustomFactor: gtsam::NoiseModelFactor {
+  // Note CustomFactor will not be wrapped for MATLAB, as there is no supporting machinery there.
```
But, it's still in `gtsam.i`. Explain (in comment) how the wrapper code knows to not wrap this and make sure MATLAB wrapper still works.

ProfFan replied:
done

dellaert commented on gtsam/nonlinear/CustomFactor.cpp:
```diff
@@ -0,0 +1,58 @@
+/*
+ * Calculates the unwhitened error by invoking the callback functor (i.e. from Python).
+ */
+Vector CustomFactor::unwhitenedError(const Values& x, boost::optional<std::vector<Matrix>&> H) const {
```
very long lines. Are you sure you're linting/formatting with Google style?

ProfFan replied:
Yes, it's still shorter than 80 :)

dellaert commented on gtsam/nonlinear/CustomFactor.h:
```diff
@@ -0,0 +1,113 @@
+  Vector unwhitenedError(const Values &x, boost::optional<std::vector<Matrix> &> H = boost::none) const override;
+
+  /** print */
+  void print(const std::string& s,
```
Can we add a cpp file to have the implementation of this?
We can then also avoid including `<iostream>` here? A bit baffled it is not included and still works, probably because NonLinearFactor is templated and it has to include it.

ProfFan replied:
Yup, I moved it anyways.

dellaert commented on python/FACTORS.md:
```diff
@@ -0,0 +1,78 @@
+# GTSAM Python-based factors
```
Rename to "CustomFactors.md"

ProfFan replied:
done

dellaert commented on python/FACTORS.md:
```diff
@@ -0,0 +1,78 @@
+
+One now can build factors purely in Python using the `CustomFactor` factor.
+
+## Theory
```
I'd move this to the end and call it "implementation", and add some more meat copied from your excellent comments in the source files

ProfFan replied:
done

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,156 @@
+        """Test the creation of a new CustomFactor"""
+
+        def error_func(this: CustomFactor, v: gtsam.Values, H: List[np.ndarray]):
+            return np.array([1, 0, 0])
```
doc! Don't you use a linter?

ProfFan replied:
This is a callback, not a class method, so no linter warnings.

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,156 @@
+        expected_pose = Pose2(1, 1, 0)
+
+        def error_func(this: CustomFactor, v: gtsam.Values, H: List[np.ndarray]) -> np.ndarray:
+            key0 = this.keys()[0]
```
doc! Don't you use a linter?

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,156 @@
+        gT2 = Pose2(-1, 4, np.pi)
+
+        def error_func(this: CustomFactor, v: gtsam.Values, _: List[np.ndarray]):
+            key0 = this.keys()[0]
```
doc! Don't you use a linter?

dellaert commented on python/gtsam/tests/test_custom_factor.py:
```diff
@@ -0,0 +1,156 @@
+        expected = Pose2(2, 2, np.pi / 2)
+
+        def error_func(this: CustomFactor, v: gtsam.Values, H: List[np.ndarray]):
+            key0 = this.keys()[0]
```
doc! Don't you use a linter?

dellaert reviewed (changes requested):
Cool ! Starting to get there.

ProfFan commented:
I have addressed all comments @dellaert  @varunagrawal

dellaert reviewed (changes requested):
After design review, need to expand example to trajopt, and think about autodiff (after that).

ProfFan commented:
Changes:
- Per @dellaert 's request, I added the example trajectory optimization with CustomFactor
- Unit test of optimizing an actual graph
- Fixed linearization of `CustomFactor` with TBB. Now it will not deadlock trying to acquire the GIL

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+from typing import List, Optional
+from functools import partial
+
+# Simulate a car for one second
```
Make a function for this that you call below.

ProfFan replied:
done

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+    :param this: gtsam.CustomFactor handle
+    :param values: gtsam.Values
+    :param jacobians: Optional list of Jacobians
+    :param measurement: GPS measurement, to be filled with `partial`
```
Can you put measurement first? That makes it like “currying” in functionali languages.

ProfFan replied:
done

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+    return error
+
+
+for k in range(5):
```
Document what’s happening here

dellaert replied:
Also, formatting?

dellaert replied:
Finally, maybe assign factor to a variable

ProfFan replied:
done

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+for k in range(5):
+    factor_graph.add(gtsam.CustomFactor(gps_model, [unknown[k]], partial(error_gps, measurement=np.array([g[k]]))))
+
+v = gtsam.Values()
```
Remove new line at end and Document block

ProfFan replied:
done

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+    v.insert(unknown[i], np.array([0.0]))
+
+params = gtsam.GaussNewtonParams()
+optimizer = gtsam.GaussNewtonOptimizer(factor_graph, v, params)
```
Same

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+
+result = optimizer.optimize()
+
+error = np.array([(result.atVector(unknown[k]) - x[k])[0] for k in range(5)])
```
What’s happening??? Document

ProfFan replied:
done

dellaert commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+    :param this: gtsam.CustomFactor handle
+    :param values: gtsam.Values
+    :param jacobians: Optional list of Jacobians
+    :param measurement: Odometry measurement, to be filled with `partial`
```
Measurement first?

ProfFan replied:
done

dellaert reviewed (changes requested):
Great improvement. Not approving yet because we need to talk about sendable

ProfFan commented on gtsam/nonlinear/NonlinearFactor.h:
```diff
@@ -134,6 +134,14 @@ class GTSAM_EXPORT NonlinearFactor: public Factor {
+   * Should the factor be evaluated in the same thread as the caller
+   * This is to enable factors that has shared states (like the Python GIL lock)
+   */
+   virtual bool sendable() const {
```
I shamelessly borrowed this concept from Rust ([here](https://doc.rust-lang.org/nomicon/send-and-sync.html)). Basically it indicates whether an object can be shared with another thread. Maybe `shareable` is a better name?

ProfFan commented on python/CustomFactors.md:
```diff
@@ -0,0 +1,110 @@
+
+## Example
+
+The following is a simple `BetweenFactor` implemented in Python.
```
It's not zero now, if you look at the last unit test.

ProfFan commented on python/gtsam/examples/CustomFactorExample.py:
```diff
@@ -0,0 +1,154 @@
+import gtsam
```
fixed

ProfFan commented:
Merged! 🤯
